### PR TITLE
feat(appsync-dart-visitor): Merge changes for amplify-flutter preview release

### PR DESCRIPTION
**Pull request by HuiSF**
**_Please Coordinate with amplify-flutter team to merge and release._**

#### Description of changes

**Update 11/30**

Feature flags `emitAuthProvider` and `isTimestampFieldsAdded` were introduce much earlier than when amplify-flutter implemented the support of them.  For most of amplify-flutter new customers, these two feature flags have been set as true by default. Therefore, checking value of these two feature flag within amplify-flutter context is outdated. 

In the latest change, enable auth provider and timestamp fields for amplify-flutter consumers by default when they are using amplify-flutter version 0.3.0 and above.

---

This PR includes changes that have been merged into the feature branch [non-model-flutter](https://github.com/aws-amplify/amplify-codegen/tree/non-model-flutter) but omitting the [last commit](https://github.com/aws-amplify/amplify-codegen/commit/06ee64a2fcbfaa37f1ae36677515f260d39e353f) which the corresponding change will be released in later version of amplify-flutter. The going to be released changes have impact only on amplify-flutter library.

Included changes:

* https://github.com/aws-amplify/amplify-codegen/pull/234
* https://github.com/aws-amplify/amplify-codegen/pull/272
* https://github.com/aws-amplify/amplify-codegen/pull/263

Related amplify-flutter changes:
* https://github.com/aws-amplify/amplify-flutter/pull/920
* https://github.com/aws-amplify/amplify-flutter/pull/1024
* https://github.com/aws-amplify/amplify-flutter/pull/955

Related amplify-android changes (for non-model type support):
* https://github.com/aws-amplify/amplify-android/pull/1459
* https://github.com/aws-amplify/amplify-android/pull/1495
* https://github.com/aws-amplify/amplify-android/pull/1513

#### Description of how you validated changes
* Added unit tests in codegen package and tests passed
* Added unit tests in amplify-flutter pacakge tests passed 
* Added unit tests in amplify-android package tests passed 
* Added integration tests in amplify-flutter package tests passed
* amplify-android integration tests passed when merging require amplify-android changes
* amplify-flutter internal manual testing/bug bash

#### Checklist

- [x] PR description included
- [x `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [x] Breaking changes to existing customers are released behind a feature flag or major version update
- [x] Changes are tested using sample applications for all relevant platforms (iOS/android/flutter/Javascript) that use the feature added/modified

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by AaronZyLee — APPROVED** [review]
The read-only field change LGTM 🐙

File: packages/amplify-codegen/src/commands/models.js [hunk 1]
```diff
@@ -118,6 +119,7 @@ async function generateModels(context) {
       enableDartNullSafety,
       handleListNullabilityTransparently,
       usePipelinedTransformer,
+      enableDartNonModelGeneration,
```
Comment (frimfram) [inline packages/amplify-codegen/src/commands/models.js:122]: checked the current preset/visitor in master branch but could not find the place where ```enableDartNonModelGeneration``` is used.

File: packages/amplify-codegen/src/commands/models.js [hunk 2]
```diff
@@ -105,7 +106,7 @@ async function generateModels(context) {
     }
   }
   const handleListNullabilityTransparently = readFeatureFlag('codegen.handleListNullabilityTransparently');
-
+  const enableDartNonModelGeneration = projectConfig.frontend === 'flutter' ? validateAmplifyFlutter(projectRoot) : false;
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/amplify-codegen/src/commands/models.js:109]: Thanks for the suggestion! I'd like suggest to reorganize the implementation of this module in a separate PR to create private util functions. Refactoring the file within in this PR seems a bit risker.

File: packages/amplify-codegen/src/commands/models.js (hunk 1 above)
Comment (HuiSF) [reply packages/amplify-codegen/src/commands/models.js:122]: Synced offline to clarify this comment.

File: packages/amplify-codegen/src/utils/validateAmplifyFlutter.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const semver = require('semver');
+
+const { PUBSPEC_FILE_NAME } = require('./validateDartSDK');
+
+function validateAmplifyFlutter(projectRoot) {
+  try {
+    const config = yaml.load(fs.readFileSync(path.join(projectRoot, PUBSPEC_FILE_NAME), 'utf8'));
+    //check dependency version
+    const version = semver.minVersion(config.dependencies.amplify_flutter);
+    if (semver.satisfies(version, '>= 0.3.0 || >= 0.3.0-rc.2')) {
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/amplify-codegen/src/utils/validateAmplifyFlutter.js]: Synced up with @AaronZyLee who implemented this util function, this function serves a single purpose to check this fixed version number. Hence renamed the function to clarify its purpose and keeping this hardcoded version number as it's not used anywhere else or for other use cases.

File: packages/amplify-codegen/src/utils/validateAmplifyFlutter.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      return true;
+    }
+    return false;
+  } catch (e) {
+    return false;
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/amplify-codegen/src/utils/validateAmplifyFlutter.js]: Added logging following suggestions from @AaronZyLee . Added unit tests to verify this flow as well.

**Comment by HuiSF** [comment]
Not sure why the some of the generator sanpshot tests are failing. This PR doesn't change anything that may have an impact on those tests. Should be caused by merging master into the PR branch. Taking a look.

**Review by AaronZyLee — APPROVED** [review]

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts [hunk 5]
```diff
@@ -237,6 +285,37 @@ export class AppSyncModelDartVisitor<
...
+    const classDeclarationBlock = new DartDeclarationBlock()
+      .asKind('class')
+      .withName(this.getModelName(model))
+      .withComment(`This is an auto generated class representing the ${model.name} type in your schema.`)
+      .annotate(['immutable']);
+    //model fields
+    model.fields.forEach(field => {
+      this.generateModelField(field, '', classDeclarationBlock);
+    });
+    //getters
```
Comment (lazpavel) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:300]: nit: these comments don't add to much value, the intent is already in the method names

**Review by lazpavel — APPROVED** [review]
LGTM

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts (hunk 5 above)
Comment (HuiSF) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:300]: Thanks and agreed. I followed existing code pattern when implemented this method... I can sync up with codegen maintainers and help cleaning up separately.

**Review by AaronZyLee — APPROVED** [review]

**Review by dnys1 — APPROVED** [review]

**Review by AaronZyLee — APPROVED** [review]

**Pull request merged**